src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,317 @@
+        };
+
+        serialized_release.@foreach ((key, val) => {
+            switch (key) {
+                case "Filename":
+                    release.filename = val.get_string ();
+                    break;
+                case "Name":
+                    release.name = val.get_string ();
+                    break;
+                case "Summary":
+                    release.summary = val.get_string ();
+                    break;
+                case "Version":
+                    release.version = val.get_string ();
+                    break;
+                case "Description":
+                    release.description = val.get_string ()
+                    .replace ("<p>", "")
+                    .replace ("</p>", "\n\n")
+                    .replace ("<li>", " • ")
+                    .replace ("</li>", "\n")
+                    .replace ("<ul>", "")
+                    .replace ("</ul>", "\n")
+                    .replace ("<ol>", "") // TODO: add support for ordered lists
+                    .replace ("</ol>", "\n")
+                    .strip ();
+                    break;
+                case "Protocol":
+                    release.protocol = val.get_string ();
+                    break;
+                case "RemoteId":
+                    release.remote_id = val.get_string ();
+                    break;
+                case "AppstreamId":
+                    release.appstream_id = val.get_string ();
+                    break;
+                case "Checksum":
+                    release.checksum = val.get_string ();
+                    break;
+                case "Vendor":
+                    release.vendor = val.get_string ();
+                    break;
+                case "Size":
+                    release.size = val.get_uint64 ();
+                    break;
+                case "License":
+                    release.license = val.get_string ();
+                    break;
+                case "TrustFlags":
+                    release.flag = (ReleaseFlag) val.get_uint64 ();
+                    break;
+                case "InstallDuration":
+                    release.install_duration = val.get_uint32 ();
+                    break;
+                case "Uri":
+                    release.uri = val.get_string ();
+                    break;
+                default:
+                    break;
+            }
+        });
+
+        return release;
+    }
+
+    private string get_path (string uri) {
+        var parts = uri.split ("/");
+        string file_path = parts[parts.length - 1];
+        return Path.build_path (Path.DIR_SEPARATOR_S, Environment.get_tmp_dir (), file_path);
+    }
+
+    public async string? download_file (Device device, string uri) {
+        var path = get_path (uri);
+
+        File server_file = File.new_for_uri (uri);
+        File local_file = File.new_for_path (path);
+
+        if (FileUtils.test (path, FileTest.IS_REGULAR)) {
```

Comment:
I'm not sure I like this. What if we're downloading a file with a really common name like `image.jpg` and something else has already created that file in the tmp directory? We would return completely the wrong file. It's better to just remove this statement and let the `FileCopyFlags.OVERWRITE` deal with it.

Reply:
Or if you efficiently want to check that it's the same file and not download it again, you could check the timestamps like what the screenshot cache in AppCenter does.

Reply:
I will address this on a follow up ticket. Fwupd provides checksums for the files.